core/base/dim.hpp

```diff
@@ -0,0 +1,229 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright 2017-2018
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_BASE_DIM_HPP_
+#define GKO_CORE_BASE_DIM_HPP_
+
+
+#include "core/base/types.hpp"
+#include "core/base/utils.hpp"
+
+
+namespace gko {
+
+
+/**
+ * A type representing the dimensions of a multidimensional object.
+ *
+ * @tparam Dimensionality  number of dimensions of the object
+ * @tparam DimensionType  datatype used to represent each dimension
+ */
+template <size_type Dimensionality, typename DimensionType = size_type>
+struct dim {
+    static constexpr size_type dimensionality = Dimensionality;
+
+    using dimension_type = DimensionType;
+
+    /**
+     * Creates a dimension object with all dimensions set to the same value.
+     *
+     * @param size  the size of each dimension
+     */
+    constexpr GKO_ATTRIBUTES dim(const dimension_type &size = dimension_type{})
+        : first_{size}, rest_{size}
+    {}
+
+    /**
+     * Creates a dimension object with the specified dimensions.
+     *
+     * If the number of dimensions given is less than the dimensionality of the
+     * object, the remaining dimensions are set to the same value as the last
+     * value given.
+     *
+     * @param first  first dimension
+     * @param rest  other dimensions
```

Comment:
maybe a little more specific? Matrices might be the most common use type...

Reply:
Good point, I've added an example for matrices to the constructor and `operator []`.